### ISSUE TITLE
UsedInMeasures and UsedInCalculatedColumns

### ISSUE THREAD

**Issue by Gezahegne** [otykier/TabularEditor#112]
I am wondering if there were UsedInMeasures and UsedInCalculatedColumns like UsedInRelationships and UsedInHierarchies. I want to do the following in the Best Practice Analyzer:
1. Get the list of hidden measures that are not used within other measures.
2. Get the list of hidden columns that are not used with other measures or Calculated Columns.

Thanks,
Gez

**Comment by otykier** [comment]
Hi,
This will be possibly with the next release, that introduces two new properties: “DependsOn” and “ReferencedBy”. I will provide the rule expressions as well, since this is indeed what the properties are useful for.

**Comment by kalmenk** [comment]
Great to hear. Thanks Daniel! Would be nice to set up a list of best practices that we can post and be shared/downloaded by others to optimize and standardize their models.

**Comment by otykier** [comment]
Check out this repo: https://github.com/TabularEditor/BestPracticeRules

**Comment by kalmenk** [comment]
Nice! @Gezahegne FYI

**Comment by Gezahegne** [comment]
Thanks Daniel for setting up the platform for sharing best practice snippets.

**Comment by otykier** [comment]
Just a quick update on this for future reference. 2.7 now exposes the `DependsOn` property on all objects that have DAX expressions (measures, calculated columns, calculated tables) and the `ReferencedBy` property on all objects that can be referenced in other DAX expressions (well, (except for KPIs)[https://github.com/otykier/TabularEditor/issues/134]).

The `DependsOn` property is an object of type [DependsOnList](https://github.com/otykier/TabularEditor/blob/9dcbba152c727b815ebb84a2d9041a4536f3e3e4/TOMWrapper/Utils/DependencyList.cs#L14), while the `ReferencedBy` property is an object of type [ReferencedByList](https://github.com/otykier/TabularEditor/blob/9dcbba152c727b815ebb84a2d9041a4536f3e3e4/TOMWrapper/Utils/DependencyList.cs#L151). Both these objects support the following properties, making it easy to detect if an object references / is referenced by other objects:

- `Count`: The number of direct dependencies / directly dependant objects
- `Measures`: A list of measures that directly depend on / are referenced by the current object
- `Columns`: A list of (calculated) columns that directly depend on / are referenced by the current object
- `Tables`: A list of (calculated) tables that directly depend on / are referenced by the current object.

In short, this allows to create, for example, a BPA rule that detects hidden columns which are not referenced by anything, by using the following Dynamic LINQ expression in the BPA rule:

```
IsHidden and ReferencedBy.Count = 0
```

**Comment by marcosqlbi** [comment]
Well done!

**Comment by otykier** [comment]
I should note that this uses the existing DAX dependency detection mechanism in Tabular Editor, which is not perfect (especially if you don't follow the [DAX best practices](https://www.sqlbi.com/tv/dax-best-practices/)), so there is still a chance that a BPA rule such as the one above, might report a column as being not used, even though it actually is in use, or vice versa.

**Comment by Gezahegne** [comment]
Thanks a lot Daniel for implementing these properties! They will help a lot in implementing BPA rules.
